app/src/main/java/jaelson/wagner/silva/upcomingmovies/ui/movies/upcoming/UpcomingMoviesPresenter.java

```diff
@@ -0,0 +1,155 @@
+package jaelson.wagner.silva.upcomingmovies.ui.movies.upcoming;
+
+import android.support.annotation.NonNull;
+import android.util.SparseArray;
+import io.reactivex.Observable;
+import io.reactivex.Observer;
+import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.disposables.Disposable;
+import io.reactivex.schedulers.Schedulers;
+import jaelson.wagner.silva.upcomingmovies.ApplicationConfiguration;
+import jaelson.wagner.silva.upcomingmovies.common.BasePresenter;
+import jaelson.wagner.silva.upcomingmovies.data.model.movies.Genre;
+import jaelson.wagner.silva.upcomingmovies.data.model.movies.Movie;
+import jaelson.wagner.silva.upcomingmovies.data.model.movies.MoviesGenresResult;
+import jaelson.wagner.silva.upcomingmovies.data.model.movies.UpcomingMoviesResult;
+import jaelson.wagner.silva.upcomingmovies.service.TheMovieDbRepository;
+
+import javax.inject.Inject;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Created by jaelsonwagner@gmail.com.
+ */
+
+public class UpcomingMoviesPresenter
+    extends BasePresenter<UpcomingMoviesContract.View>
+    implements UpcomingMoviesContract.Presenter {
+
+  private TheMovieDbRepository repository;
+  private UpcomingMoviesContract.View view;
+  private SparseArray<String> moviesGenresMapping = new SparseArray<>();
```

Comment:
Shouldn't this be cached for more than the single instance of this fragment? Should the presenter even be aware of this?

Reply:
Sorry! This point it's not clear for me. Can you explain more?

Reply:
I was suggesting that this could be `static`, or better yet, used only on the `Repository`. The presenter would receive a list of movies with their genres, and wouldn't even need to know there is a different endpoint for that.

Reply:
I disagree use static, since the remote data can change. But, I agree that this SparseArray could be delivery by the repository. At first moment, I did not know that I should transforms the received genres. So, this solution can be improved as you said. Well noted.